adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -1196,12 +1201,12 @@ protected final void close(final MqttEndpoint endpoint, final Device authenticat
 
         final Span span = newSpan("CLOSE", endpoint, authenticatedDevice, traceSamplingPriority);
         onClose(endpoint);
-        cmdHandler.removeAllSubscriptions((tenant, device) -> {
-            final Span closeHandlerSpan = newSpan("Send Disconnected Event", endpoint, authenticatedDevice,
-                    traceSamplingPriority);
-            sendDisconnectedTtdEvent(tenant, device, authenticatedDevice, span.context())
-                    .setHandler(sendAttempt -> closeHandlerSpan.finish());
-        });
+        final CompositeFuture removalDoneFuture = cmdHandler.removeAllSubscriptions((tenant, device) -> {
+            final Span sendEventSpan = newChildSpan(span.context(), "Send Disconnected Event", endpoint,
+                    authenticatedDevice);
+            return sendDisconnectedTtdEvent(tenant, device, authenticatedDevice, sendEventSpan.context())
+                    .onComplete(r -> sendEventSpan.finish()).mapEmpty();
```

Comment:
If I am not mistaken, then this code sends a disconnect event *for each* subscription being canceled. However, FMPOV we only need to send the disconnect *once after all* subscriptions have been canceled. WDYT?

Reply:
IMHO, the only scenario, where an MQTT device would want to create multiple subscriptions over the same MQTT connection, would be for a gateway device to add multiple subscriptions for commands targeted at different, specific devices. In that case, the `sendDisconnectedTtdEvent` method with its `deviceId` parameter should actually be called for _all_ these devices separately.
So, I think, the code makes sense, the way it is.